### PR TITLE
[JENKINS-59638] Notify Testers

### PR DESCRIPTION
**Pull request by championofblocks**
- Adding boolean flag for notifyTesters to main API call site
- Updating tests

### CODE REVIEW

**Comment by championofblocks** [comment]
Hi @mezpahlan 👋 
Sure thing I'll open a ticket for this and fix the compilation issues.

**Comment by championofblocks** [comment]
@mezpahlan ticket is open here: https://issues.jenkins-ci.org/browse/JENKINS-59638

still working on the code

File: src/main/java/io/jenkins/plugins/appcenter/AppCenterRecorder.java [hunk 1]
```diff
@@ -99,6 +102,10 @@ public String getReleaseNotes() {
         return Util.fixNull(releaseNotes);
     }
 
+    public String getNotifyTesters() {
+        return Boolean.toString(notifyTesters);
```
Comment (mezpahlan) [inline src/main/java/io/jenkins/plugins/appcenter/AppCenterRecorder.java]: Why does this need to be converted to a String only to be converted back again to a boolean later?

File: README.md [hunk 2]
```diff
@@ -5,7 +5,7 @@ plugin.
 
 ## Roadmap
 
-This plugin is currently in Alpha and looking for contributors. To begin with it will aim to support the upload
```
Comment (mezpahlan) [inline README.md]: What's the deal with these changes? Are the necessary in this PR? I am of the opinion they are not.

File: README.md [hunk 3]
```diff
@@ -39,10 +39,11 @@ stage('Publish') {
             ownerName: 'janes-addiction',
             appName: 'ritual-de-lo-habitual',
             pathToApp: 'three/days/xiola.apk',
-            distributionGroups: 'casey, niccoli'
+            distributionGroups: 'casey, niccoli',
+            notifyTesters: true
```
Comment (mezpahlan) [inline README.md]: I wanted to limit this example to the simplest form of using the plugin. In which case this flag would default to true as discussed.

I think you're right in that we should provide examples of other usages but can we do that elsewhere please.

File: src/test/java/io/jenkins/plugins/appcenter/task/internal/DistributeResourceTaskTest.java [hunk 4]
```diff
@@ -81,4 +81,4 @@ public void should_ReturnException_When_RequestIsUnSuccessful() {
         assertThat(exception).hasCauseThat().hasCauseThat().isInstanceOf(HttpException.class);
         assertThat(exception).hasCauseThat().hasCauseThat().hasMessageThat().isEqualTo("HTTP 500 Server Error");
     }
-}
\ No newline at end of file
+}
```
Comment (mezpahlan) [inline src/test/java/io/jenkins/plugins/appcenter/task/internal/DistributeResourceTaskTest.java]: As well as these tests there is probably also a specific test that you can write to assert the correct notify testers flag is sent to the server depending on what the user has selected.

File: src/main/java/io/jenkins/plugins/appcenter/AppCenterRecorder.java (hunk 1 above)
Comment (championofblocks) [reply src/main/java/io/jenkins/plugins/appcenter/AppCenterRecorder.java]: The `envVars.expand()` in `UploadRequest` only takes a string. Can you recommend a better way? Still getting to know the code here and my Java is rusty 😝

File: README.md (hunk 2 above)
Comment (championofblocks) [reply README.md]: Agreed. This is Resharper suggestions kicking in for me here. I'll revert the unnecessary changes here

File: README.md (hunk 3 above)
Comment (championofblocks) [reply README.md]: Sure thing. I'll revert this

File: src/test/java/io/jenkins/plugins/appcenter/task/internal/DistributeResourceTaskTest.java (hunk 4 above)
Comment (championofblocks) [reply src/test/java/io/jenkins/plugins/appcenter/task/internal/DistributeResourceTaskTest.java]: For sure. I'll try and add that

File: src/main/java/io/jenkins/plugins/appcenter/AppCenterRecorder.java (hunk 1 above)
Comment (mezpahlan) [reply src/main/java/io/jenkins/plugins/appcenter/AppCenterRecorder.java]: Ahh, OK. I see. In that case I feel that not everything needs to have an environment variable expansion. So keep it as a boolean  defaulted to true and pass that straight on.

**Comment by mezpahlan** [comment]
Hi @championofblocks I'm going to jump on this PR so that it moves along. You might see some commits from me. Remember to pull before you commit.

**Comment by championofblocks** [comment]
@mezpahlan no problem. I wish I had more time to commit to it at the moment 😄

File: README.md (hunk 3 above)
Comment (mezpahlan) [reply README.md]: I've raised [JENKINS-59692](https://issues.jenkins-ci.org/browse/JENKINS-59692) to fill in examples.

File: src/main/java/io/jenkins/plugins/appcenter/AppCenterRecorder.java (hunk 1 above)
Comment (mezpahlan) [reply src/main/java/io/jenkins/plugins/appcenter/AppCenterRecorder.java]: Whilst the other fields are more naturally represented as strings and it makes sense to offer variable expansion (Jenkins users love variable expansions) for this field I don't think it does. I've changed this to simply pass through the configured value which will default to `true` if not explicitly set to `false.

File: src/test/java/io/jenkins/plugins/appcenter/task/internal/DistributeResourceTaskTest.java (hunk 4 above)
Comment (mezpahlan) [reply src/test/java/io/jenkins/plugins/appcenter/task/internal/DistributeResourceTaskTest.java]: I've added these now.

**Comment by mezpahlan** [comment]
@championofblocks Could you give this a quick test please before I merge this into master. Let me know if it works for you or not. You should be able to get the artefacts from [here](https://ci.jenkins.io/blue/organizations/jenkins/Plugins%2Fappcenter-plugin/detail/PR-27/9/artifacts).

**Comment by mezpahlan** [comment]
Well I have tested this myself and successfully received a notification from AppCenter so I'm going to go ahead and merge this. Anything comes up, raise a new issue please. Thanks.

**Pull request merged**